pkg/process/events/listener_linux.go

```diff
@@ -0,0 +1,177 @@
+type SysProbeListener struct {
+	// client holds a gRPC client to connect to system-probe
+	client api.SecurityModuleClient
+	// conn holds the connection used by the client, so it can be closed once the listener is stopped
+	conn *grpc.ClientConn
+	// retryInterval is how long the listener will wait before trying to reconnect to system-probe if there's a connection failure
+	retryInterval time.Duration
+	// handler is the EventHandler function applied to every event collected by the listener
+	handler EventHandler
+	// connected holds the status of the connection to system-probe
+	connected atomic.Value
+	// wg and exit are used to control the main listener routine
+	wg   sync.WaitGroup
+	exit chan struct{}
+}
+
+// NewListener returns a new SysProbeListener to listen for process events
+func NewListener(handler EventHandler) (*SysProbeListener, error) {
+	socketPath := ddconfig.Datadog.GetString("runtime_security_config.socket")
+	if socketPath == "" {
+		return nil, errors.New("runtime_security_config.socket must be set")
+	}
+
+	conn, err := grpc.Dial(socketPath, grpc.WithInsecure(), grpc.WithContextDialer(func(ctx context.Context, url string) (net.Conn, error) {
+		return net.Dial("unix", url)
+	}))
+	if err != nil {
+		return nil, err
+	}
+
+	client := api.NewSecurityModuleClient(conn)
+	return newSysProbeListener(conn, client, handler)
+}
+
+// newSysProbeListener returns a new SysPobeListener
+func newSysProbeListener(conn *grpc.ClientConn, client api.SecurityModuleClient, handler EventHandler) (*SysProbeListener, error) {
+	if handler == nil {
+		return nil, errors.New("can't create a Listener without an EventHandler")
+	}
+
+	return &SysProbeListener{
+		client:        client,
+		conn:          conn,
+		retryInterval: 2 * time.Second,
+		handler:       handler,
+		exit:          make(chan struct{}),
+	}, nil
+}
+
+// Run starts a new thread to listen for process events
+func (l *SysProbeListener) Run() {
+	log.Info("Start listening for process events")
+
+	l.wg.Add(1)
+	go func() {
+		defer l.wg.Done()
+		l.run()
+	}()
+}
+
+// run keeps polling the SecurityModule server for process events
+func (l *SysProbeListener) run() {
+	l.connected.Store(false)
+	logTicker := newLogBackoffTicker()
+
+	for {
+		select {
+		case <-l.exit:
+			return
+		default:
+			stream, err := l.client.GetProcessEvents(context.Background(), &api.GetProcessEventParams{})
+			if err != nil {
+				l.connected.Store(false)
+
+				select {
+				case <-logTicker.C:
+					log.Warnf("Error while connecting to the runtime-security module: %v", err)
+				default:
+					// do nothing
+				}
+
+				time.Sleep(l.retryInterval)
+				continue
+			}
+
+			if l.connected.Load() != true {
+				l.connected.Store(true)
+
+				log.Info("Successfully connected to the runtime-security module")
+			}
+
+			for {
+				in, err := stream.Recv()
+				if err == io.EOF || in == nil {
+					break
+				}
+				l.consumeData(in.Data)
+			}
```

Comment:
Do we need a check here for the exit channel?

Reply:
Oh, thanks for catching it!
There's a check for the exit channel in the external loop but in this case the listener have to wait until all messages are sent by the system-probe before exiting.
I've updated it to add a check in the stream loop as well :https://github.com/DataDog/datadog-agent/pull/12099/commits/e8d1135918624a8dbd0cec8887434eb3688978c2.